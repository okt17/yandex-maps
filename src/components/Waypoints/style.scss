.app__waypoints {
  margin: 0.75rem;

  &__input {
    color: white !important;
    width: 100%;
  }
}
