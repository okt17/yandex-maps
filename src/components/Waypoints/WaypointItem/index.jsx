import React from 'react';
import PropTypes from 'prop-types';
import { IconButton, Icon } from '@material-ui/core';
import './style';

const BCN = 'app__waypoints__item'; // Base Class Name

const toFixed = n => n.toFixed(3);

class WaypointItem extends React.PureComponent {
  static propTypes = {
    deleteWaypoint: PropTypes.func.isRequired,
    index: PropTypes.number.isRequired,
    waypoint: PropTypes.object.isRequired,
  };
  handleDeleteButtonClick = () => this.props.deleteWaypoint(this.props.waypoint.id);
  handleDragStart = (event) => {
    event.dataTransfer.effectAllowed = 'move';
    event.dataTransfer.setData('waypoint', this.props.index);

    return true;
  };
  render() {
    const {
      waypoint: {
        name,
        coordinates,
      },
    } = this.props;

    return (
      <div
        className={BCN}
        draggable
        onDragStart={this.handleDragStart}
      >
        <span className={`${BCN}__container`}>
          <div>{name}</div>

          <div>{coordinates.map(toFixed).join(', ')}</div>
        </span>

        <IconButton
          color='secondary'
          onClick={this.handleDeleteButtonClick}
          size='small'
          title='Удалить точку маршрута'
        >
          <Icon>clear</Icon>
        </IconButton>
      </div>
    );
  }
}

export default WaypointItem;
