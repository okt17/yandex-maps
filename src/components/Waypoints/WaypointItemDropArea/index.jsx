import React from 'react';
import PropTypes from 'prop-types';
import './style';

const BCN = 'app__waypoints__drop-area'; // Base Class Name

class WaypointItemDropArea extends React.PureComponent {
  state = {};
  static propTypes = {
    index: PropTypes.number.isRequired,
    moveWaypoint: PropTypes.func.isRequired,
  };
  handleDrop = (event) => {
    event.preventDefault();

    const index = parseInt(event.dataTransfer.getData('waypoint'), 10);

    if (Number.isNaN(index) === false) {
      this.props.moveWaypoint(index, this.props.index);
    }

    if (this.state.active) {
      this.setState({ active: false });
    }
  };
  handleDragEnter = (event) => {
    event.preventDefault();

    if (event.dataTransfer.items[0].type === 'waypoint') {
      this.setState({ active: true });

      return true;
    }

    return false;
  };
  handleDragLeave = (event) => {
    event.preventDefault();

    if (this.state.active) {
      this.setState({ active: false });
    }
  };
  handleDragOver = (event) => {
    event.preventDefault();
  };
  render() {
    const { active } = this.state;

    return (
      <div
        className={`${BCN}${active ? ` ${BCN}_active` : ''}`}
        onDrop={this.handleDrop}
        onDragEnter={this.handleDragEnter}
        onDragLeave={this.handleDragLeave}
        onDragOver={this.handleDragOver}
      />
    );
  }
}

export default WaypointItemDropArea;
