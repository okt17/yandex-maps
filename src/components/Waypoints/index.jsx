import React from 'react';
import PropTypes from 'prop-types';
import { Input } from '@material-ui/core';
import WaypointItem from './WaypointItem';
import WaypointItemDropArea from './WaypointItemDropArea';
import './style';

class Waypoints extends React.Component {
  static propTypes = {
    addWaypoint: PropTypes.func.isRequired,
    deleteWaypoint: WaypointItem.propTypes.deleteWaypoint,
    moveWaypoint: WaypointItemDropArea.propTypes.moveWaypoint,
    newWaypointName: PropTypes.string,
    setNewWaypointName: PropTypes.func.isRequired,
    waypoints: PropTypes.array.isRequired,
  };
  handleInputChange = event => this.props.setNewWaypointName(event.target.value);
  handleInputFocus = (event) => {
    event.target.select();
  };
  handleInputKeydown = (event) => {
    // Enter
    if (event.keyCode === 13) {
      this.props.addWaypoint();

      // focus the input after handling the event
      const input = event.target;
      // put this on top of the event loop queue
      setTimeout(() => input.select(), 0);
    }
  };
  render() {
    const {
      deleteWaypoint,
      moveWaypoint,
      newWaypointName,
      waypoints,
    } = this.props;

    return (
      <div className='app__waypoints'>
        <Input
          autoFocus
          className='app__waypoints__input'
          onChange={this.handleInputChange}
          onFocus={this.handleInputFocus}
          onKeyDown={this.handleInputKeydown}
          value={newWaypointName}
        />

        {
          waypoints.map((waypoint, i) => (
            <React.Fragment key={waypoint.id}>
              <WaypointItemDropArea
                index={i}
                moveWaypoint={moveWaypoint}
              />

              <WaypointItem
                index={i}
                waypoint={waypoint}
                deleteWaypoint={deleteWaypoint}
              />
            </React.Fragment>
          ))
        }

        <WaypointItemDropArea
          index={waypoints.length}
          moveWaypoint={moveWaypoint}
        />
      </div>
    );
  }
}

export default Waypoints;
