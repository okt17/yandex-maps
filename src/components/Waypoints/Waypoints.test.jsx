/* eslint-disable no-undef */
import React from 'react';
import renderer from 'react-test-renderer';
import Waypoints from '.';

function pass() {}

test('Waypoints component test', () => {
  const component = renderer.create(<Waypoints
    addWaypoint={pass}
    deleteWaypoint={pass}
    moveWaypoint={pass}
    newWaypointName='value'
    setNewWaypointName={pass}
    waypoints={[
      {
        id: '1',
        name: 'waypoint1',
        coordinates: [20, 18],
      },
    ]}
  />);

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
