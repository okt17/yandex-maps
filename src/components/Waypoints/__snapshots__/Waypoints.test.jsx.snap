// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Waypoints component test 1`] = `
<div
  className="app__waypoints"
>
  <div
    className="MuiInput-root-1 MuiInput-underline-5 app__waypoints__input"
  >
    <input
      autoFocus={true}
      className="MuiInput-input-9"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      type="text"
      value="value"
    />
  </div>
  <div
    className="app__waypoints__drop-area"
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
  />
  <div
    className="app__waypoints__item"
    draggable={true}
    onDragStart={[Function]}
  >
    <span
      className="app__waypoints__item__container"
    >
      <div>
        waypoint1
      </div>
      <div>
        20.000, 18.000
      </div>
    </span>
    <button
      className="MuiButtonBase-root-20 MuiIconButton-root-14 MuiIconButton-colorSecondary-17"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      size="small"
      tabIndex="0"
      title="Удалить точку маршрута"
      type="button"
    >
      <span
        className="MuiIconButton-label-19"
      >
        <span
          aria-hidden="true"
          className="material-icons MuiIcon-root-23"
        >
          clear
        </span>
      </span>
      <span
        className="MuiTouchRipple-root-29"
      />
    </button>
  </div>
  <div
    className="app__waypoints__drop-area"
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
  />
</div>
`;
